runners/spark/src/main/java/org/apache/beam/runners/spark/io/SourceRDD.java

```diff
@@ -0,0 +1,201 @@
+                                                               TaskContext context) {
+      final Iterator<WindowedValue<T>> iter = new Iterator<WindowedValue<T>>() {
+        @SuppressWarnings("unchecked")
+        SourcePartition<T> partition = (SourcePartition<T>) split;
+        BoundedSource.BoundedReader<T> reader = createReader(partition);
+
+        private boolean finished = false;
+        private boolean started = false;
+        private boolean closed = false;
+
+        @Override
+        public boolean hasNext() {
+          try {
+            if (!started) {
+              started = true;
+              finished = !reader.start();
+            } else {
+              finished = !reader.advance();
+            }
+            if (finished) {
+              // safely close the reader if there are no more elements left to read.
+              closeIfNotClosed();
+            }
+            return !finished;
+          } catch (IOException e) {
+            closeIfNotClosed();
+            throw new RuntimeException("Failed to read from reader.", e);
+          }
+        }
+
+        @Override
+        public WindowedValue<T> next() {
+          return WindowedValue.timestampedValueInGlobalWindow(reader.getCurrent(),
+              reader.getCurrentTimestamp());
+        }
+
+        @Override
+        public void remove() {
+          throw new UnsupportedOperationException("Remove from partition iterator is not allowed.");
+        }
+
+        private void closeIfNotClosed() {
+          if (!closed) {
+            closed = true;
+            try {
+              reader.close();
+            } catch (IOException e) {
+              throw new RuntimeException("Failed to close Reader.", e);
+            }
+          }
+        }
+      };
+
+      return new InterruptibleIterator<>(context,
+          scala.collection.JavaConversions.asScalaIterator(iter));
+    }
+
+    private BoundedSource.BoundedReader<T> createReader(SourcePartition<T> partition) {
+      try {
+        return ((BoundedSource<T>) partition.source).createReader(
+            runtimeContext.getPipelineOptions());
+      } catch (IOException e) {
+        throw new RuntimeException("Failed to create reader from a BoundedSource.", e);
+      }
+    }
+  }
+
+  /**
+   * An input {@link Partition} wrapping the partitioned {@link Source}.
+   */
+  private static class SourcePartition<T> implements Partition {
+
+    private final int rddId;
+    private final int index;
+    private final Source<T> source;
+
+    SourcePartition(int rddId, int index, Source<T> source) {
+      this.rddId = rddId;
+      this.index = index;
+      this.source = source;
+    }
+
+    @Override
+    public int index() {
+      return index;
+    }
+
+    @Override
+    public int hashCode() {
+      return 41 * (41 + rddId) + index;
+    }
+
+    @Override
+    public boolean equals(Object other) {
+      return super.equals(other);
```

Comment:
According to your implementation of `hashCode` this should be
```
if (!(other instanceof SourcePartition)) {
  return false;
}
SourcePartition that = (SourcePartition) other;
return this.rddId == that.rddId && this.index == that.index
```

Which probably also means that you should include the `Source` in the equals and hash code methods